common/src/tap/checks/value_check.rs

```diff
@@ -0,0 +1,535 @@
+            .get_expected_value(agora_query)
+            .map_err(CheckError::Failed)?;
+
+        // get value
+        let value = receipt.signed_receipt().message.value;
+
+        let should_accept = value >= expected_value;
+
+        tracing::trace!(
+            value,
+            expected_value,
+            should_accept,
+            "Evaluating mininum query fee."
+        );
+
+        if should_accept {
+            Ok(())
+        } else {
+            return Err(CheckError::Failed(anyhow!(
+                "Query receipt does not have the minimum value. Expected value: {}. Received value: {}.",
+                expected_value, value,
+            )));
+        }
+    }
+}
+
+fn compile_cost_model(model: String, variables: String) -> anyhow::Result<CostModel> {
+    if model.len() > (1 << 16) {
+        return Err(anyhow!("CostModelTooLarge"));
+    }
+    let model = CostModel::compile(&model, &variables)?;
+    Ok(model)
+}
+
+#[derive(serde::Deserialize)]
+#[serde(tag = "tg_op")]
+enum CostModelNotification {
+    #[serde(rename = "INSERT")]
+    Insert {
+        deployment: String,
+        model: String,
+        variables: String,
+    },
+    #[serde(rename = "DELETE")]
+    Delete { deployment: String },
+}
+
+#[cfg(test)]
+mod tests {
+    use alloy::primitives::{address, Address};
+    use std::time::Duration;
+
+    use sqlx::PgPool;
+    use tap_core::receipt::{checks::Check, Context, ReceiptWithState};
+    use tokio::time::sleep;
+
+    use crate::{
+        cost_model::test::{add_cost_models, global_cost_model, to_db_models},
+        tap::AgoraQuery,
+        test_vectors::create_signed_receipt,
+    };
+
+    use super::MinimumValue;
+
+    #[sqlx::test(migrations = "../migrations")]
+    async fn initialize_check(pgpool: PgPool) {
+        let check = MinimumValue::new(pgpool).await;
+        assert_eq!(check.cost_model_map.read().unwrap().len(), 0);
+    }
+
+    #[sqlx::test(migrations = "../migrations")]
+    async fn should_initialize_check_with_models(pgpool: PgPool) {
+        // insert 2 cost models for different deployment_id
+        let test_models = crate::cost_model::test::test_data();
+
+        add_cost_models(&pgpool, to_db_models(test_models.clone())).await;
+
+        let check = MinimumValue::new(pgpool).await;
+        assert_eq!(check.cost_model_map.read().unwrap().len(), 2);
+
+        // no global model
+        assert!(check.global_model.read().unwrap().is_none());
+    }
+
+    #[sqlx::test(migrations = "../migrations")]
+    async fn should_watch_model_insert(pgpool: PgPool) {
+        let check = MinimumValue::new(pgpool.clone()).await;
+        assert_eq!(check.cost_model_map.read().unwrap().len(), 0);
+
+        // insert 2 cost models for different deployment_id
+        let test_models = crate::cost_model::test::test_data();
+        add_cost_models(&pgpool, to_db_models(test_models.clone())).await;
+        sleep(Duration::from_millis(200)).await;
```

Comment:
The `sleep` makes things a bit confusing.

You could use something like this:

```rust
async fn wait_for_update(check: &MinimumValue) {
    tokio::time::timeout(Duration::from_secs(1), async {
        while check.cost_model_map.read().unwrap().len() != expected_len {
            sleep(Duration::from_millis(50)).await;
        }
    }).await.unwrap();
}
```

I think the intention is clearer.